stan/math/opencl/rev/vari.hpp

```diff
@@ -0,0 +1,190 @@
+#ifndef STAN_MATH_OPENCL_REV_VARI_HPP
+#define STAN_MATH_OPENCL_REV_VARI_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/rev/core/vari.hpp>
+#include <stan/math/opencl/is_matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * The variable implementation for `matrix_cl`.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, require_kernel_expression_lhs_t<T>>
+    : public vari_base, public chainable_alloc {
```

Comment:
Listing this as very optional, but we could do a map/unmap and use the stack allocators memory like something from here. Though idt we need to worry about that right now

https://software.intel.com/content/www/us/en/develop/articles/getting-the-most-from-opencl-12-how-to-increase-performance-by-minimizing-buffer-copies-on-intel-processor-graphics.html

Reply:
That is benefitial only if running OpenCL kernels on CPU or integrated GPUs that use host memory anyway. For discrete GPUs copies would happen anyway during the kernel runtime (and they could possibly be finer grained making them slower).

Reply:
Yes thats why I'm commenting it as very optional. Could be neat! But not necessary at all

Reply:
Than we would have to detect whether we are running on discrete GPU or not and manage memory differently. This is definitely possible, but I am not sure if it is necessary. I doubt many peopre will run OpenCL on integrated GPUs or CPUs. My guess is that only people who have discrete GPU and want to use it even try using OpenCL. Anyway if we decide to do it, it will definitely happen in its own PR.